src/control/server/ctl_storage_rpc.go

```diff
@@ -242,68 +242,80 @@ func (c *ControlService) scmFormat(scmCfg storage.ScmConfig, reformat bool) (*ct
 
 // doFormat performs format on storage subsystems, populates response results
 // in storage subsystem routines and broadcasts (closes channel) if successful.
-func (c *ControlService) doFormat(i *IOServerInstance, reformat bool, resp *ctlpb.StorageFormatResp) error {
+func (c *ControlService) doFormat(srv *IOServerInstance, reformat bool) (resp *ctlpb.StorageFormatResp) {
+	resp = new(ctlpb.StorageFormatResp)
+	resp.Mrets = proto.ScmMountResults{}
+	resp.Crets = proto.NvmeControllerResults{}
+
+	srvIdx := srv.Index()
 	needsSuperblock := true
 	needsScmFormat := reformat
 	skipNvmeResult := newCret(c.log, "format", "", ctlpb.ResponseStatus_CTL_SUCCESS, "",
-		fmt.Sprintf(msgNvmeFormatSkip, i.Index()))
+		fmt.Sprintf(msgNvmeFormatSkip, srv.Index()))
 
 	c.log.Infof("formatting storage for %s instance %d (reformat: %t)",
-		DataPlaneName, i.Index(), reformat)
+		DataPlaneName, srv.Index(), reformat)
 
-	scmConfig := i.scmConfig()
+	scmConfig := srv.scmConfig()
 
-	// If not reformatting, check if SCM is already formatted.
-	if !reformat {
+	scmErrored := false
+	scmErr := func(err error) *ctlpb.ScmMountResult {
+		scmErrored = true
+		return newMntRet(c.log, "format", scmConfig.MountPoint,
+			ctlpb.ResponseStatus_CTL_ERR_SCM, err.Error(),
+			fault.ShowResolutionFor(err))
+	}
+
+	if srv.IsStarted() {
+		resp.Mrets = append(resp.Mrets, scmErr(errors.Errorf(
+			"instance %d: can't format storage of running instance", srvIdx)))
```

Comment:
Just a suggestion - I think returning early might be a little clearer than tracking if we've seen an error via a bool at key points. If you refactor to make it possible to defer appending the skipNvmeResult, it will be possible to do that at least. Or alternately if you break doScmFormat and doNvmeFormat into separate functions, the top-level caller can add the skipNvmeResult if doScmFormat reported failure.

Reply:
done